Improve performance of block dragging.  This is a backport of the blo…

picklesrus opened the pull request:
…ck drag surface from scratch-blocks.  At the beginning of a block drag, blocks get moved to a drag surface which then translates using translate3d to avoid repainting the entire svg on every mouse move.  At the end of the drag, the blocks are dropped back in the svg in their new position.

NeilFraser commented on core/block_svg.js:
```diff
@@ -342,15 +364,75 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
+ */
+Blockly.BlockSvg.prototype.translate = function(x, y) {
+  this.getSvgRoot().setAttribute('transform',
+    'translate(' + x + ',' + y + ')');
```
More indent.

NeilFraser commented on core/block_svg.js:
```diff
@@ -342,15 +364,75 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
+ */
+Blockly.BlockSvg.prototype.clearTransformAttributes_ = function() {
+   if (this.getSvgRoot().hasAttribute('transform')) {
+      this.getSvgRoot().removeAttribute('transform');
```
Less indent.  Here and next line.

NeilFraser commented on core/block_svg.js:
```diff
@@ -342,15 +364,75 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
+      this.getSvgRoot().removeAttribute('transform');
+   }
+  if (this.getSvgRoot().hasAttribute('style')) {
+    this.getSvgRoot().removeAttribute('style');
```
Attempting to remove an attribute that is not on the element doesn't raise an exception and may be faster than checking.  Here and above.

NeilFraser commented on core/blockly.js:
```diff
@@ -125,6 +125,13 @@ Blockly.dragMode_ = Blockly.DRAG_NONE;
 
 /**
+ * Cached value for whether 3D is supported
+ * @type {boolean}
```
!boolean, since it starts null.
Better yet, set it false here, then overwrite?

picklesrus replied:
I need to be able to tell the difference between has been checked and hasn't because is3dSupported check is expensive. It is cacheable though, thus the initialization to null as an indicator that we haven't run the check yet.

NeilFraser commented on core/css.js:
```diff
@@ -147,6 +147,7 @@ Blockly.Css.CONTENT = [
   '.injectionDiv {',
     'height: 100%;',
     'position: relative;',
+    'overflow:hidden', /* So blocks in drag surface disappear at edges */
```
Space after colon, and also needs semicolon.

NeilFraser commented on core/css.js:
```diff
@@ -447,6 +463,7 @@ Blockly.Css.CONTENT = [
     'overflow-x: visible;',
     'overflow-y: auto;',
     'position: absolute;',
+    'z-index:70;',
```
Space needed.

NeilFraser commented on core/inject.js:
```diff
@@ -27,6 +27,7 @@
 goog.provide('Blockly.inject');
 
 goog.require('Blockly.Css');
+goog.require('Blockly.BlockDragSurfaceSvg');
```
Should be alphabetical.

NeilFraser commented on core/utils.js:
```diff
@@ -719,3 +739,43 @@ Blockly.utils.wrapToText_ = function(words, wordBreaks) {
+  }
+  // CC-BY-SA Lorenzo Polidori
+  // https://stackoverflow.com/questions/5661671/detecting-transform-translate3d-support
+  if (!window.getComputedStyle) {
```
Use goog.global instead of window.  NodeJS and non-browser environments don't have window.

NeilFraser commented on core/utils.js:
```diff
@@ -719,3 +739,43 @@ Blockly.utils.wrapToText_ = function(words, wordBreaks) {
+  }
+
+  document.body.removeChild(el);
+  Blockly.cache3dSupported_ = (has3d !== undefined && has3d.length > 0 && has3d !== "none");
```
Suggestion:
Blockly.cache3dSupported_ = !!(has3d && has3d !== 'none');

NeilFraser commented on core/utils.js:
```diff
@@ -719,3 +739,43 @@ Blockly.utils.wrapToText_ = function(words, wordBreaks) {
+    };
+
+  // Add it to the body to get the computed style.
+  document.body.insertBefore(el, null);
```
Here's an underused function you might like:
https://developer.mozilla.org/en-US/docs/Web/API/Node/replaceChild

picklesrus replied:
Oh neat. Thanks. Not sure it is applicable here since this isn't really replacing anything.  It is inserting the note into the dom to check whether it can perform certain 3d stuff, then removing it.

NeilFraser replied:
Ah, my mistake.  Didn't read the code carefully enough.

NeilFraser commented on core/workspace_svg.js:
```diff
@@ -48,18 +48,24 @@ goog.require('goog.userAgent');
+ 
+ if (opt_blockDragSurface) {
+    this.blockDragSurface_ = opt_blockDragSurface;
+ 
```
Unneeded blank line.

picklesrus commented:
Thanks. And sorry I forgot to run lint. Fixed a few more things lint found as well. Think I got 'em all.

Also edge is now fixed, but this approach is slower in Edge so we may decide to disable it.

picklesrus commented:
Did a little more performance testing in Edge. It is fairly bad both for the existing code and with this change.  With one or two blocks, this definitely makes it worse. With 100+, they're both terrible. And with 10-20, they're pretty similarly bad.

All testing done on Surface 3 so a pretty low end device. For consistency it would be nice to have this change on everywhere.  My vote is to check this in with edge every once in a while and see if the performance of diverges.

tmickel commented on core/block_svg.js:
```diff
@@ -342,15 +364,71 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
+ * @private
+ */
+Blockly.BlockSvg.prototype.clearTransformAttributes_ = function() {
+  this.getSvgRoot().removeAttribute('transform');
```
I forget, did we decide that this is unnecessary, or not?

When we move blocks to the drag surface, I think they should maintain their `transform` attribute, not switch to `style` (as was a previous implementation). If that's the case, then I think we can drop this function and the calls to it.

picklesrus replied:
The style one isn't necessary. The transform once is.  This is the transform is the one on the block itself, which is part of what getRelativeToSurfaceXY returns.  If we don't clear it, we end up adding it in twice. I guess we could change getRelativeToSurfaceXY, except that it is used in a bunch more places.  Instead I guess we could add another helper that doesn't take into account the transform instead?  But, clearing it here instead seems simpler to me.

tmickel commented on core/css.js:
```diff
@@ -156,6 +157,21 @@ Blockly.Css.CONTENT = [
+    '-webkit-backface-visibility: hidden;',
+    'backface-visibility: hidden;',
+    '-webkit-perspective: 1000;',
+    'perspective: 1000;',
```
I would double check, but I'm not confident whether these CSS properties (backface visibility, perspective) are necessary or just left-over from experiments.

picklesrus replied:
Ah, good call. Removed. Don't see a difference.

picklesrus commented on core/block_drag_surface.js:
```diff
@@ -0,0 +1,182 @@
+ */
+Blockly.BlockDragSurfaceSvg.prototype.translateAndScaleGroup = function(x, y, scale) {
+  this.scale_ = scale;
+  // Force values to have two decimal points.
```
Fixed

picklesrus commented on core/utils.js:
```diff
@@ -266,6 +274,18 @@ Blockly.getRelativeXY_ = function(element) {
   }
+
+    // Third, check for style="transform: translate3d(...)".
+  var style = element.getAttribute('style');
```
Fixed (I think).

rachel-fenichel commented on core/block_drag_surface.js:
```diff
@@ -0,0 +1,180 @@
+ */
+
+/**
+ * @fileoverview Functions for injecting Blockly into a web page.
```
@fileoverview is wrong.

rachel-fenichel commented on core/block_drag_surface.js:
```diff
@@ -0,0 +1,180 @@
+
+
+/**
+ * Class for a Drag Surface SVG.
```
Class for a drag surface for the currently dragged block.  This is a separate SVG that contains only the currently moving block, or nothing.

rachel-fenichel commented on core/block_drag_surface.js:
```diff
@@ -0,0 +1,180 @@
+ * @constructor
+ */
+Blockly.BlockDragSurfaceSvg = function(container) {
+  this.container_ = container;
```
@type tag here, and @private.

rachel-fenichel commented on core/block_drag_surface.js:
```diff
@@ -0,0 +1,180 @@
+Blockly.BlockDragSurfaceSvg.prototype.SVG_ = null;
+
+/**
+ * SVG group inside the drag surface. This is where blocks are moved to.
```
This is where blocks live while they are being dragged.

tmickel replied:
if the drag surface is enabled :)

rachel-fenichel commented on core/block_drag_surface.js:
```diff
@@ -0,0 +1,180 @@
+
+/**
+ * Set the SVG blocks on the drag surface's group and show the surface.
+ * Only one block should be on the drag surface at a time.
```
Only one block group should be on the drag surface at a time.

rachel-fenichel commented on core/block_drag_surface.js:
```diff
@@ -0,0 +1,180 @@
+ * @param {number} y Y translation
+ * @param {number} scale Scale of the group.
+ */
+Blockly.BlockDragSurfaceSvg.prototype.translateAndScaleGroup = function(x, y, scale) {
```
If this should only be called at the beginning or end of a drag, include that in the comment.

tmickel replied:
if the switch below is made, I think it doesn't need to be called at the end of a drag, only the beginning.

rachel-fenichel commented on core/block_svg.js:
```diff
@@ -342,15 +364,70 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
+
+/**
+ * Move this block back to the workspace block canvas.
+ * Generally should be called at the same time as setDragging_(false).
```
Does nothing if useDragSurface_ is false.

rachel-fenichel commented on core/css.js:
```diff
@@ -156,6 +157,17 @@ Blockly.Css.CONTENT = [
+    'right: 0;',
+    'bottom: 0;',
+    'overflow: visible !important;',
+    'z-index: 50;', /* Display below toolbox, but everything else. */
```
Display below toolbox, but above everything else.

rachel-fenichel commented on core/utils.js:
```diff
@@ -238,6 +238,14 @@ Blockly.isTargetInput_ = function(e) {
          e.target.type == 'tel' || e.target.type == 'url' ||
          e.target.isContentEditable;
 };
+/**
```
Newline above this.

rachel-fenichel commented on core/workspace_svg.js:
```diff
@@ -509,6 +522,9 @@ Blockly.WorkspaceSvg.prototype.translate = function(x, y) {
   this.svgBlockCanvas_.setAttribute('transform', translation);
   this.svgBubbleCanvas_.setAttribute('transform', translation);
+  if (this.blockDragSurface_) {
+    this.blockDragSurface_.translateAndScaleGroup(x, y, this.scale);
```
Does this need to be updated every time the workspace translates?

I guess the advantage is that we don't have this slowing down block drag starts, but doesn't it make every workspace movement slower?

tmickel replied:
I believe it basically makes no difference. You could move it to the block drag start, which would add some trivial amount of time there - but should be insignificant compared to the repaint which is about to happen. OTOH, calling it whenever the workspace translates and scales should also be very fast, since the element is hidden and there should be nothing in the SVG. I originally put it here because it made most sense to me conceptually as the goal is to match the workspace translation/scale.

rachel-fenichel replied:
Sweet--if it's fast because it's hidden, then that's fine.

tmickel commented:
LGTM!!

picklesrus commented:
Sweet. I'll merge once the develop->master merge (PR #747) lands.

rachel-fenichel commented:
https://i.imgur.com/eMqtr39.gif
